explorer/parse_and_execute/parse_and_execute.cpp

```diff
@@ -23,16 +23,18 @@ static auto PrintTimingOnExit(TraceStream* trace_stream, const char* label,
   auto end = std::chrono::steady_clock::now();
   auto duration = end - *cursor;
   *cursor = end;
-
-  return llvm::make_scope_exit([=]() {
+  auto exit_scope_function = llvm::make_scope_exit([=]() {
+    trace_stream->set_current_phase(ProgramPhase::Timing);
     if (trace_stream->is_enabled()) {
       *trace_stream << "Time elapsed in " << label << ": "
                     << std::chrono::duration_cast<std::chrono::milliseconds>(
                            duration)
                            .count()
                     << "ms\n";
+      trace_stream->set_current_phase(ProgramPhase::Unknown);
```

Comment:
There are a few instances of this pattern in this file in particular:
1) set the current phase
2) do some tracing
3) restore the old phase

I think it would be better for step 3 to restore the old phase instead, and it would be good to add an [RAII type](https://en.cppreference.com/w/cpp/language/raii) to do that. I'm happy to skip that in this patch and do it as a separate PR, though; we've already got some good progress here that would be great to land.